build_container/build_container_mobile.sh

```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
+set -e
+
+./build_container_ubuntu.sh
```

Comment:
Do we want to just start the base image from the Ubuntu one? That saves time of running this script

Reply:
Would it actually save time though?

Because the `build_container_linux mobile` and `build_container_linux ubuntu` CI jobs run in parallel with the current setup, overall CI completes faster than if we waited for `build_container_linux ubuntu` to complete and then ran `build_container_linux mobile` pulling from what was just built.

In fact, the mobile job only builds for one architecture so it finishes faster than the ubuntu one (28 min vs 38 min), so currently this should really increase the CI wall time, assuming there's no increased queueing.

Reply:
If you use staged build (within same Dockerfile and mark it mobile) instead of spawning multiple CI jobs, it shouldn't add extra time to CI (max of amd64 ubuntu + mobile, arm64 ubuntu).

That way the images built would share layers so it saves storage and transfer.

Reply:
+1 re making use of docker layers

Reply:
Ah ok, got it! I’ll try that.

Reply:
The Dockerfile part seems pretty straightforward:

```dockerfile
# envoy-build-ubuntu
FROM ubuntu:focal AS ubuntu

# Workaround for: https://issuetracker.google.com/issues/216325949
ENV CLOUDSDK_PYTHON=python3

COPY ./build_container_common.sh /
COPY ./build_container_ubuntu.sh /

RUN ./build_container_ubuntu.sh

ENV LANG en_US.utf8

# envoy-build-mobile
FROM ubuntu AS mobile
COPY ./build_container_mobile.sh /

ENV ANDROID_SDK_INSTALL_TARGET /.android
ENV ANDROID_HOME /.android/sdk
ENV ANDROID_SDK_ROOT /.android/sdk
ENV ANDROID_NDK_VERSION 21.4.7075529
ENV ANDROID_NDK_HOME /.android/sdk/ndk/21.4.7075529

RUN ./build_container_mobile.sh
```

And I can build the multiplatform `ubuntu` image with `--target ubuntu` and then another amd64-only build for `mobile` with `--target mobile` which will reuse the layers from the amd64 ubuntu image.

I'm struggling to find a nice way to update `docker_build_linux.sh` though. Especially while trying to achieve what @lizan said in his comment, to build `amd64 ubuntu + mobile` in parallel with `arm64 ubuntu` in the same CI job.

In `azure-pipelines.yml` I'm thinking I'd define some targets to build like this:

```yaml
OS_DISTRO: ubuntu
PUSH_GCR_IMAGE: true
GCR_IMAGE_NAME: envoy-build
DOCKER_TARGETS: "ubuntu mobile"
```

But then I'm unclear on how I'd adjust the lines that build, tag and push in `docker_build_linux.sh`:

```bash
ci_log_run docker buildx build . -f "Dockerfile-${OS_DISTRO}" -t "${IMAGE_NAME}:${CONTAINER_TAG}" --platform "${BUILD_TOOLS_PLATFORMS}"

for IMAGE_TAG in "${IMAGE_TAGS[@]}"; do
  ci_log_run docker buildx build . -f "Dockerfile-${OS_DISTRO}" -t "${IMAGE_TAG}" --platform "${BUILD_TOOLS_PLATFORMS}" --push
done
```